source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,416 @@
+    if (!entry->active()) {
+      stats_.lb_subsets_active_.dec();
+      stats_.lb_subsets_removed_.inc();
+    }
+  }
+}
+
+bool SubsetLoadBalancer::hostMatchesDefaultSubset(const Host& host) {
+  const envoy::api::v2::Metadata& host_metadata = host.metadata();
+
+  for (const auto& it : default_subset_.fields()) {
+    const ProtobufWkt::Value& host_value = Config::Metadata::metadataValue(
+        host_metadata, Config::MetadataFilters::get().ENVOY_LB, it.first);
+
+    if (!ValueUtil::equal(host_value, it.second)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool SubsetLoadBalancer::hostMatches(const SubsetMetadata& kvs, const Host& host) {
+  const envoy::api::v2::Metadata& host_metadata = host.metadata();
+
+  for (const auto& kv : kvs) {
+    const ProtobufWkt::Value& host_value = Config::Metadata::metadataValue(
+        host_metadata, Config::MetadataFilters::get().ENVOY_LB, kv.first);
+
+    if (!ValueUtil::equal(host_value, kv.second)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Iterates over subset_keys looking up values from the given host's
+// metadata. Each key-value pair is appended to kvs. Returns a
+// non-empty value if the host has a value for each key.
+SubsetLoadBalancer::SubsetMetadata
+SubsetLoadBalancer::extractSubsetMetadata(const std::set<std::string>& subset_keys,
+                                          const Host& host) {
+  SubsetMetadata kvs;
+
+  const envoy::api::v2::Metadata& metadata = host.metadata();
+  const auto& filter_it = metadata.filter_metadata().find(Config::MetadataFilters::get().ENVOY_LB);
+  if (filter_it == metadata.filter_metadata().end()) {
+    return kvs;
+  }
+
+  const auto& fields = filter_it->second.fields();
+  for (const auto key : subset_keys) {
+    const auto it = fields.find(key);
+    if (it == fields.end()) {
+      break;
+    }
+    kvs.emplace_back(std::pair<std::string, ProtobufWkt::Value>(key, it->second));
+  }
+
+  if (kvs.size() != subset_keys.size()) {
+    kvs.clear();
+  }
+
+  return kvs;
+}
+
+// Given a vector of key-values (from extractSubsetMetadata),
+// recursively finds the matching LbSubsetEntryPtr.
+SubsetLoadBalancer::LbSubsetEntryPtr
+SubsetLoadBalancer::findOrCreateSubset(LbSubsetMap& subsets, const SubsetMetadata& kvs,
+                                       uint32_t idx) {
+  ASSERT(idx < kvs.size());
+
+  const std::string& name = kvs[idx].first;
+  const ProtobufWkt::Value& pb_value = kvs[idx].second;
+  const HashedValue value(pb_value);
+
+  const auto& kv_it = subsets.find(name);
+  if (kv_it != subsets.end()) {
+    ValueSubsetMap& value_subset_map = kv_it->second;
+    const auto vs_it = value_subset_map.find(value);
+    if (vs_it != value_subset_map.end()) {
+      LbSubsetEntryPtr entry = vs_it->second;
+      idx++;
+      if (idx == kvs.size()) {
+        return entry;
+      }
+
+      return findOrCreateSubset(entry->children_, kvs, idx);
+    }
+  }
+
+  // not found, but don't want creation
```

Comment:
nit: start with capital, period end of sentence (same elsewhere if applicable, haven't been paying full attention)